src/Base/Unit.cpp

```diff
@@ -35,24 +38,31 @@
 using namespace Base;
 
 // clang-format off
-static inline void checkPow(UnitSignature sig, double exp)
-{
-    auto isInt = [](double value) {
-        return std::fabs(std::round(value) - value) < std::numeric_limits<double>::epsilon();
+constexpr int UnitSignatureLengthBits                   = 4;
+constexpr int UnitSignatureMassBits                     = 4;
+constexpr int UnitSignatureTimeBits                     = 4;
+constexpr int UnitSignatureElectricCurrentBits          = 4;
+constexpr int UnitSignatureThermodynamicTemperatureBits = 4;
+constexpr int UnitSignatureAmountOfSubstanceBits        = 4;
+constexpr int UnitSignatureLuminousIntensityBits        = 4;
+constexpr int UnitSignatureAngleBits                    = 4;
+
+union UnitSignature {
+    uint32_t Value;
+    struct {
```

Comment:
This is not strict ISO C++ but a GNU extension. Because for my local build I use `-Werror` this gives me a build failure: `FreeCAD/src/Base/Unit.cpp:52:5: error: anonymous structs are a GNU extension [-Werror,-Wgnu-anonymous-struct]`

Reply:
The warning must be ignored with:
```

union UnitSignature {
    uint32_t Value;
#if defined(__clang__)
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wgnu-anonymous-struct"
#elif defined(__GNUC__)
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wgnu-anonymous-struct"
#endif
    struct {
        int32_t Length: UnitSignatureLengthBits;
        int32_t Mass: UnitSignatureMassBits;
        int32_t Time: UnitSignatureTimeBits;
        int32_t ElectricCurrent: UnitSignatureElectricCurrentBits;
        int32_t ThermodynamicTemperature: UnitSignatureThermodynamicTemperatureBits;
        int32_t AmountOfSubstance: UnitSignatureAmountOfSubstanceBits;
        int32_t LuminousIntensity: UnitSignatureLuminousIntensityBits;
        int32_t Angle: UnitSignatureAngleBits;
    };
#if defined(__clang__)
#pragma clang diagnostic pop
#elif defined(__GNUC__)
#pragma GCC diagnostic pop
#endif
};
```

Reply:
According to https://dev.to/pauljlucas/unions-stdvariant-in-c-2op1 the whole union is not legal C++ code and can lead to UB (see the section about type punning).

Reply:
Ok. So would you prefer `memcpy` or `std::bit_cast`? I'm temped to use the latter.

Reply:
Hmm, `std::bit_cast` is C++20 thing.